File: PiBenchmark.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * PiBenchmark
 * * A simple, cross-platform Java CPU benchmark tool that calculates Pi
 * to a specified number of digits using multiple threads.
 * * This tool is designed to be run from the command line to test CPU performance.
 * * How to compile:
 * javac PiBenchmark.java
 * * How to run:
 * java PiBenchmark [digits] [repsPerThread] [threads]
 * * Arguments:
 * [digits]:        The number of decimal places to calculate Pi (e.g., 5000)
 * [repsPerThread]: The number of repetitions EACH thread will run (e.g., 1000)
 * [threads]:       The number of concurrent threads to use (e.g., 8)
 * * Example:
 * java PiBenchmark 5000 1000 8
 */
public class PiBenchmark {

    // These constants are used for the Machin-like formula for Pi
    // Pi = 16 * arctan(1/5) - 4 * arctan(1/239)
    private static final BigDecimal FOUR = new BigDecimal("4");
    private static final BigDecimal SIXTEEN = new BigDecimal("16");
    private static final BigDecimal BD_5 = new BigDecimal("5");
    private static final BigDecimal BD_239 = new BigDecimal("239");

    public static void main(String[] args) {
        
        // --- 1. Argument Parsing ---
        if (args.length != 3) {
            System.err.println("Usage: java PiBenchmark [digits] [repsPerThread] [threads]");
            System.err.println("Example: java PiBenchmark 5000 1000 8");
            return;
        }

        int digits = 0;
        int repsPerThread = 0;
        int numThreads = 0;

        try {
            digits = Integer.parseInt(args[0]);
            repsPerThread = Integer.parseInt(args[1]);
            numThreads = Integer.parseInt(args[2]);

            if (digits <= 0 || repsPerThread <= 0 || numThreads <= 0) {
                throw new NumberFormatException("Inputs must be positive integers.");
            }
        } catch (NumberFormatException e) {
            System.err.println("Invalid arguments. All inputs must be positive integers.");
            System.err.println(e.getMessage());
            return;
        }

        // --- 2. Warm-up Phase ---
        // This is crucial to allow the JVM's JIT compiler to optimize the code
        // before we start the timed benchmark.
        // We run a smaller number of calculations to "warm up" the critical methods.
        int warmUpReps = (int)Math.max(1, repsPerThread * 0.1); // 10% of reps, or at least 1
        for (int i = 0; i < warmUpReps; i++) {
            calculatePi(digits);
        }

        // --- 3. Timed Benchmark ---
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        long startTime = System.nanoTime();

        // Create and submit a task for each thread
        // Each task will run the Pi calculation 'repsPerThread' times
        for (int i = 0; i < numThreads; i++) {
            final int taskDigits = digits; // Need final variable for lambda
            final int taskReps = repsPerThread;
            
            Runnable task = () -> {
                for (int j = 0; j < taskReps; j++) {
                    calculatePi(taskDigits);
                }
            };
            executor.submit(task);
        }

        // Wait for all threads to finish
        executor.shutdown();
        try {
            // Wait a long time, but this will return as soon as all tasks are done
            executor.awaitTermination(1, TimeUnit.HOURS);
        } catch (InterruptedException e) {
            System.err.println("Benchmark was interrupted.");
            Thread.currentThread().interrupt();
            return;
        }

        long endTime = System.nanoTime();

        // --- 4. Calculate and Report Results ---
        long totalTimeNano = endTime - startTime;
        double totalTimeMs = totalTimeNano / 1_000_000.0;
        double totalTimeSec = totalTimeNano / 1_000_000_000.0;

        long totalCalculations = (long)repsPerThread * numThreads;
        double calculationsPerSecond = totalCalculations / totalTimeSec;

        // Print the final, single-line output
        System.out.printf(
            "Test:Pi, Digits:%d, RepsPerThread:%d, Threads:%d, TotalTime(ms):%.2f, CalcPerSec:%.2f%n",
            digits,
            repsPerThread,
            numThreads,
            totalTimeMs,
            calculationsPerSecond
        );
    }

    /**
     * Calculates Pi to the specified number of digits using a Machin-like formula.
     * This method is the "work" for the benchmark.
     * * @param digits The number of decimal places to calculate.
     * @return A BigDecimal representing Pi.
     */
    public static BigDecimal calculatePi(int digits) {
        // We need extra precision for intermediate calculations
        int scale = digits + 10;
        
        // arctan(x) = x - x^3/3 + x^5/5 - x^7/7 + ...
        // We use the formula: Pi = 16 * arctan(1/5) - 4 * arctan(1/239)
        BigDecimal term1 = arctan(BD_5.pow(-1), scale);
        BigDecimal term2 = arctan(BD_239.pow(-1), scale);
        
        // 16 * term1
        BigDecimal pi = term1.multiply(SIXTEEN);
        // pi = pi - (4 * term2)
        pi = pi.subtract(term2.multiply(FOUR));
        
        // Set the final scale to the requested number of digits
        return pi.setScale(digits, RoundingMode.HALF_UP);
    }

    /**
     * Calculates arctan(x) using the Taylor series expansion.
     * * @param x     The value (as BigDecimal)
     * @param scale The precision required
     * @return arctan(x)
     */
    private static BigDecimal arctan(BigDecimal x, int scale) {
        BigDecimal result = x;
        BigDecimal xSquared = x.multiply(x);
        BigDecimal term = x;
        BigDecimal divisor = BigDecimal.ONE;
        
        // Loop until the term is too small to affect the result
        while (true) {
            divisor = divisor.add(BigDecimal.valueOf(2)); // 3, 5, 7, ...
            term = term.multiply(xSquared).negate(); // -x^3, +x^5, -x^7, ...
            
            // Add the new term: term / divisor
            BigDecimal termToAdd = term.divide(divisor, scale, RoundingMode.HALF_EVEN);
            
            // If the new term is 0, we've reached the limit of our precision
            if (termToAdd.signum() == 0) {
                break;
            }
            result = result.add(termToAdd);
        }
        return result;
    }
}

File: PiBenchmark.py
#!/usr/bin/env python3

"""
PiBenchmark (Python Version)

A simple, cross-platform Python CPU benchmark tool that calculates Pi
to a specified number of digits using multiple processes.

This tool is designed to be run from the command line to test CPU performance.

How to run:
python3 pi_benchmark.py [digits] [repsPerThread] [threads]

Arguments:
[digits]:        The number of decimal places to calculate Pi (e.g., 5000)
[repsPerThread]: The number of repetitions EACH process will run (e.g., 1000)
[threads]:       The number of concurrent processes to use (e.g., 8)

Example:
python3 pi_benchmark.py 5000 1000 8

NOTE: This script uses 'multiprocessing' instead of 'threading' to bypass
the Python Global Interpreter Lock (GIL) and achieve true parallel
CPU computation. The '[threads]' argument is mapped to the number of processes.
"""

import sys
import time
from decimal import Decimal, getcontext
import multiprocessing

# --- Pi Calculation Logic (Machin-like formula) ---
# This section replicates the logic from the Java version using Python's
# 'decimal' module, which is equivalent to 'BigDecimal'.

def arctan(x_inv_str, scale):
    """
    Calculates arctan(1/x) using the Taylor series expansion.
    """
    getcontext().prec = scale
    x_inv = Decimal(x_inv_str)
    x = Decimal(1) / x_inv
    x_squared = x * x
    
    result = x
    term = x
    divisor = Decimal(1)
    
    while True:
        divisor += 2
        term *= x_squared * Decimal(-1)
        term_to_add = term / divisor
        
        # Check if the term is too small to affect the result at this precision
        # This is the Python equivalent of the Java 'signum() == 0' check
        # in a high-precision context.
        if abs(term_to_add) < Decimal('1e-%d' % scale):
            break
        result += term_to_add
        
    return result

def calculate_pi(digits):
    """
    Calculates Pi to the specified number of digits.
    This is the "work" function for the benchmark.
    
    Uses the formula: Pi = 16 * arctan(1/5) - 4 * arctan(1/239)
    """
    # We need extra precision for intermediate calculations
    scale = digits + 10
    
    # Set precision for the arctan calculations
    getcontext().prec = scale

    term1 = arctan('5', scale)
    term2 = arctan('239', scale)

    # pi = 16 * term1 - 4 * term2
    pi = (term1 * 16) - (term2 * 4)
    
    # Set the final precision
    getcontext().prec = digits
    # The `+pi` operation applies the final precision context
    return +pi

# --- Main Benchmark Execution ---

def main():
    """
    Main function to parse args, run warm-up, and execute the benchmark.
    """
    
    # --- 1. Argument Parsing ---
    if len(sys.argv) != 4:
        print("Usage: python3 pi_benchmark.py [digits] [repsPerThread] [threads]", file=sys.stderr)
        print("Example: python3 pi_benchmark.py 5000 1000 8", file=sys.stderr)
        sys.exit(1)

    try:
        digits = int(sys.argv[1])
        reps_per_thread = int(sys.argv[2])
        num_processes = int(sys.argv[3]) # This is the '[threads]' arg

        if digits <= 0 or reps_per_thread <= 0 or num_processes <= 0:
            raise ValueError("Inputs must be positive integers.")
            
    except ValueError as e:
        print(f"Invalid arguments. All inputs must be positive integers.", file=sys.stderr)
        print(f"{e}", file=sys.stderr)
        sys.exit(1)

    # --- 2. Warm-up Phase ---
    # This is still important in Python to ensure any initial setup,
    # imports, and CPU caches are "warm" before timing.
    warm_up_reps = max(1, int(reps_per_thread * 0.1)) # 10% of reps
    for _ in range(warm_up_reps):
        calculate_pi(digits)

    # --- 3. Timed Benchmark ---
    
    # Total calculations to be performed across all processes
    total_calculations = reps_per_thread * num_processes
    
    # Create a list of 'digits' arguments to map to the pool
    # The pool will distribute these tasks among the processes
    tasks = [digits] * total_calculations

    # Start the high-precision timer
    start_time = time.perf_counter()

    # Create a process pool and map the work
    # The 'with' statement automatically handles pool.close() and pool.join()
    with multiprocessing.Pool(processes=num_processes) as pool:
        # map() blocks until all tasks are complete
        pool.map(calculate_pi, tasks)

    end_time = time.perf_counter()

    # --- 4. Calculate and Report Results ---
    total_time_sec = end_time - start_time
    total_time_ms = total_time_sec * 1000.0
    
    calculations_per_second = total_calculations / total_time_sec

    # Print the final, single-line output in the exact same format
    print(
        f"Test:Pi, Digits:{digits}, RepsPerThread:{reps_per_thread}, Threads:{num_processes}, "
        f"TotalTime(ms):{total_time_ms:.2f}, CalcPerSec:{calculations_per_second:.2f}"
    )

if __name__ == "__main__":
    # This check is MANDATORY for 'multiprocessing' to work
    # correctly on all platforms (especially Windows).
    # It ensures that the 'main()' function only runs in the
    # parent process, not in the child processes.
    main()
  
